### PR TITLE
Links refactor

### PR DESCRIPTION
**Pull request by nnhw**

### CODE REVIEW

File: include/communication/models/communication.h [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+    QString type() const;
+
+    bool isConnected() const;
+
+    int bytesReceived() const;
+    int bytesSent() const;
+
+public slots:
+    void Connect();
```
Comment (MishkaRogachev) [inline include/communication/models/communication.h]: connect, disconect

File: include/communication/models/i_communication_protocol.h [hunk 2]
```diff
@@ -0,0 +1,28 @@
+#ifndef I_COMMUNICATION_PROTOCOL_H
+#define I_COMMUNICATION_PROTOCOL_H
+
+#include "QByteArray"
```
Comment (MishkaRogachev) [inline include/communication/models/i_communication_protocol.h]: <> for external libs

File: include/communication/models/link_specification.h [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+#include <QVariantMap>
+
+namespace md::domain
+{
+class LinkSpecification : public QObject
+{
+    Q_OBJECT
+
+public:
+    explicit LinkSpecification() = default;
```
Comment (MishkaRogachev) [inline include/communication/models/link_specification.h]: no explicit

File: include/communication/models/i_communication_protocol.h [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+#include "QObject"
+
+namespace md::domain
+{
+class ICommunicationProtocol : public QObject
```
Comment (MishkaRogachev) [inline include/communication/models/i_communication_protocol.h:9]: Looks like interface for data_source class

File: include/communication/models/communication.h [hunk 5]
```diff
@@ -0,0 +1,56 @@
+#ifndef COMMUNICATION_H
+#define COMMUNICATION_H
+
+#include "i_link_transceiver.h"
+#include "link_specification.h"
+#include "link_traits.h"
+#include "named.h"
```
Comment (MishkaRogachev) [inline include/communication/models/communication.h:5]: unneeded headers

File: include/communication/models/i_link_transceiver.h [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+#include <QObject>
+
+namespace md::domain
+{
+class ILinkTransceiver : public QObject
+{
+    Q_OBJECT
+
+public:
+    ILinkTransceiver(QObject* parent) : QObject(parent)
```
Comment (MishkaRogachev) [inline include/communication/models/i_link_transceiver.h:14]: Looks like interface for data_source class

File: include/communication/models/link_specification.h [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    explicit LinkSpecification(QVariantMap parameters, QObject* parent = nullptr);
```
Comment (MishkaRogachev) [inline include/communication/models/link_specification.h]: const QVariantMap&

File: include/communication/models/protocol_description.h [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+#include <QString>
+
+namespace md::domain
+{
+class ProtocolDescription : public Named
+{
+    Q_OBJECT
+
+public:
+    explicit ProtocolDescription() = default;
```
Comment (MishkaRogachev) [inline include/communication/models/protocol_description.h]: no explicit

File: include/communication/models/protocol_description.h [hunk 9]
```diff
@@ -0,0 +1,31 @@
+#ifndef PROTOCOL_DESCRIPTION_H
+#define PROTOCOL_DESCRIPTION_H
+
+#include "i_communication_protocol.h"
+#include "named.h"
+
+#include <QObject>
```
Comment (MishkaRogachev) [inline include/communication/models/protocol_description.h]: unneeded

File: include/communication/models/protocol_description.h [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+{
+class ProtocolDescription : public Named
+{
+    Q_OBJECT
+
+public:
+    explicit ProtocolDescription() = default;
+    explicit ProtocolDescription(ICommunicationProtocol* communicationProtocol,
+                                 const QString& name = QString(), QObject* parent = nullptr);
+    explicit ProtocolDescription(const ProtocolDescription& another, QObject* parent = nullptr);
```
Comment (MishkaRogachev) [inline include/communication/models/protocol_description.h]: Is copy-constructor and operator = used ?

File: include/communication/models/protocol_description.h [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+
+    ProtocolDescription& operator=(const ProtocolDescription& another);
+
+    ICommunicationProtocol* protocol() const;
+
+private:
+    ICommunicationProtocol* m_communicationProtocol;
```
Comment (MishkaRogachev) [inline include/communication/models/protocol_description.h]: This dependency breaks ability to serialize ProtocolDescription

File: include/communication/service/communication_service.h [hunk 12]
```diff
@@ -0,0 +1,42 @@
+#ifndef COMMUNICATION_SERVICE_H
+#define COMMUNICATION_SERVICE_H
+
+#include <QObject>
```
Comment (MishkaRogachev) [inline include/communication/service/communication_service.h:4]: extra headers

File: include/communication/service/communication_service.h [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+class CommunicationService
+    : public QObject
+    , public md::domain::IService
+{
+    Q_OBJECT
+
+public:
+    explicit CommunicationService(const QString& fileName);
+
+    void registerProtocol(const QString& name, domain::ICommunicationProtocol* protocol);
```
Comment (MishkaRogachev) [inline include/communication/service/communication_service.h]: Better
`registerProtocol(const ProtocolDescription& description, domain::ICommunicationProtocol* protocol);`
or
`registerProtocol(const QString& id, domain::ICommunicationProtocol* protocol);`

File: include/communication/models/link_traits.h [hunk 14]
```diff
@@ -0,0 +1,62 @@
+#ifndef LINK_TRAITS_H
+#define LINK_TRAITS_H
```
Comment (MishkaRogachev) [inline include/communication/models/link_traits.h]: Maybe communicaztion_traits

File: include/communication/models/link_traits.h [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+
+#include <QMap>
+#include <QSharedPointer>
+#include <QString>
+#include <QVariantMap>
```
Comment (MishkaRogachev) [inline include/communication/models/link_traits.h]: #include <QVariantMap> is enought for all used types

File: src/communication/models/communication.cpp [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+                             const ProtocolDescription& protocolDescription, const QString& name,
+                             QObject* parent) :
+    Named(name, utils::generateId(), parent),
+    m_linkSpecification(linkSpecification),
+    m_protocolDescription(protocolDescription),
+    m_connected(false),
+    m_bytesReceived(0),
+    m_bytesSent(0)
+{
+    m_linkTransceiver = createLinkTranceiver(linkSpecification);
```
Comment (MishkaRogachev) [inline src/communication/models/communication.cpp]: better to unamed namespace

File: src/communication/models/link_specification.cpp [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+{
+    return m_parameters.value(parameter);
+}
+
+QVariantMap LinkSpecification::parameters() const
+{
+    return m_parameters;
+}
+
+void LinkSpecification::fillInParameters()
```
Comment (MishkaRogachev) [inline src/communication/models/link_specification.cpp]: Better hide this to unammed namespace

File: src/communication/models/link_specification.cpp [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+    {
+        for (const QString& parameter : link_type::parameters.values(m_type))
+        {
+            if (!m_parameters.contains(parameter))
+            {
+                m_parameters.insert(parameter, link_type::defaultValues.value(parameter));
+            }
+        }
+    }
+}
```
Comment (MishkaRogachev) [inline src/communication/models/link_specification.cpp]: no new line

File: src/communication/service/communication_service.cpp [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+constexpr char localPort[] = "local_port";
+
+} // namespace
+
+CommunicationService::CommunicationService(const QString& fileName) :
+    m_protocols(),
+    m_communications(),
+    m_source(new data_source::JsonSourceFile(fileName)),
+    m_json(m_source->read())
+
```
Comment (MishkaRogachev) [inline src/communication/service/communication_service.cpp]: extra line

File: src/communication/service/communication_service.cpp [hunk 20]
```diff
@@ -0,0 +1,69 @@
...
+#include "json_source_file.h"
+#include "link_traits.h"
+#include "link_transceiver.h"
+#include "link_transceiver_threaded.h"
+
+using namespace md::app;
+
+namespace
+{
+constexpr char type[] = "type";
```
Comment (MishkaRogachev) [inline src/communication/service/communication_service.cpp]: can be uset form traits

File: src/communication/tranceivers/link_transceiver.cpp [hunk 21]
```diff
@@ -0,0 +1,88 @@
...
+    m_factory(),
+    m_link(createLink(linkSpecification))
+
+{
+}
+
+void LinkTransceiver::start()
+{
+    m_timerId = this->startTimer(::interval);
+    receiveData();
```
Comment (MishkaRogachev) [inline src/communication/tranceivers/link_transceiver.cpp]: this->receiveData();

File: src/communication/tranceivers/link_transceiver.cpp [hunk 22]
```diff
@@ -0,0 +1,88 @@
...
+        return QObject::timerEvent(event);
+
+    m_link->checkHandlers();
+}
+
+void LinkTransceiver::receiveData()
+{
+    m_link->asyncReceive([this](const std::string& received_data) {
+        emit receivedData(QByteArray::fromStdString(received_data));
+        receiveData();
```
Comment (MishkaRogachev) [inline src/communication/tranceivers/link_transceiver.cpp]: this->

File: src/communication/tranceivers/link_transceiver.cpp [hunk 23]
```diff
@@ -0,0 +1,88 @@
...
+    });
+}
+
+void LinkTransceiver::send(const QByteArray& data)
+{
+    m_link->asyncSend(data.toStdString(), [](std::size_t size) {});
+}
+void LinkTransceiver::openLink()
```
Comment (MishkaRogachev) [inline src/communication/tranceivers/link_transceiver.cpp:60]: new line

**Review by MishkaRogachev — CHANGES_REQUESTED** [review]
Some more questions

File: include/communication/models/i_communication_protocol.h (hunk 4 above)
Comment (nnhw) [reply include/communication/models/i_communication_protocol.h:9]: Yes, it is an interface for protocols like a Mavlink. But it is a part of a domain layer due to dependency inversion

File: include/communication/models/communication.h (hunk 5 above)
Comment (nnhw) [reply include/communication/models/communication.h:5]: Only #include "link_traits.h" is unnecessary, removed it

File: include/communication/models/i_link_transceiver.h (hunk 6 above)
Comment (nnhw) [reply include/communication/models/i_link_transceiver.h:14]: Yes, similar to https://github.com/Midgrad/kjarni/pull/5#discussion_r787389421

File: include/communication/models/protocol_description.h (hunk 10 above)
Comment (nnhw) [reply include/communication/models/protocol_description.h]: Yes, link specification needs to be [assignable]( https://doc.qt.io/qt-5/containers.html#assignable-data-type)

File: include/communication/models/protocol_description.h (hunk 11 above)
Comment (nnhw) [reply include/communication/models/protocol_description.h]: I'll search for a cleaner solution when I implement SQL storage.

For now, I think that I'll be able to implement deserialization, checking if there is a registered protocol with the same name/id.

For serializing, I'll skip this member of fill it with 0/null/nullptr.

File: include/communication/models/link_traits.h (hunk 15 above)
Comment (nnhw) [reply include/communication/models/link_traits.h]: Which include is enough?

File: include/communication/service/communication_service.h (hunk 13 above)
Comment (nnhw) [reply include/communication/service/communication_service.h]: The first one is violating the purpose of `registerProtocol`, it is supposed to create `ProtocolDescription`.  

The second one is just renaming of name -> id, and protocol description is entity and it has its own id, which is confusing.

File: include/communication/models/i_communication_protocol.h (hunk 4 above)
Comment (MishkaRogachev) [reply include/communication/models/i_communication_protocol.h:9]: Receiveing and sending byte arrays is not domain logic

File: CMakeLists.txt [hunk 24]
```diff
@@ -19,9 +19,9 @@ file(GLOB_RECURSE SOURCES "include/*" "src/*")
 target_sources(${PROJECT_NAME} PRIVATE ${SOURCES})
 
 # Link with libraries
-target_link_libraries(${PROJECT_NAME} PUBLIC Qt5::Core Qt5::Sql)
+target_link_libraries(${PROJECT_NAME} PUBLIC Qt5::Core Qt5::Sql loodsman pthread)
```
Comment (MishkaRogachev) [inline CMakeLists.txt]: Can we hide pthread over pthread

File: include/communication/models/link_traits.h (hunk 15 above)
Comment (nnhw) [reply include/communication/models/link_traits.h]: QVariantMap

File: include/communication/models/communication.h [hunk 25]
```diff
@@ -0,0 +1,48 @@
...
+class Communication : public domain::ICommunication
+{
+    Q_OBJECT
+
+public:
+    Communication(const domain::LinkSpecification& linkSpecification,
+                  const domain::ProtocolSpecification& protocolDescription,
+                  ICommunicationProtocol* protocol, const QString& name = QString(),
+                  QObject* parent = nullptr);
+    QString type() const override;
```
Comment (MishkaRogachev) [inline include/communication/models/communication.h:28]: new line

File: include/communication/models/communication.h [hunk 26]
```diff
@@ -0,0 +1,48 @@
...
+public slots:
+    void setConnected(bool connected) override;
+    void start();
+    void stop();
+
+private:
+    ILinkTransceiver* m_linkTransceiver;
+    domain::LinkSpecification m_linkSpecification;
+    domain::ProtocolSpecification m_protocolDescription;
+    ICommunicationProtocol* m_protocol;
```
Comment (MishkaRogachev) [inline include/communication/models/communication.h]: const

File: include/communication/models/communication_description.h [hunk 27]
```diff
@@ -0,0 +1,39 @@
...
+
+#include "i_communication.h"
+#include "named.h"
+
+#include "link_specification.h"
+#include "protocol_specification.h"
+
+namespace md::domain
+{
+
```
Comment (MishkaRogachev) [inline include/communication/models/communication_description.h]: empty line

File: include/communication/models/communication_description.h [hunk 28]
```diff
@@ -0,0 +1,39 @@
...
+class CommunicationDescription : public ICommunication
+{
+public:
+    CommunicationDescription(const LinkSpecification& specification,
+                             const ProtocolSpecification& protocolDescription,
+                             const QString& name = QString(), QObject* parent = nullptr);
+    QString type() const override;
```
Comment (MishkaRogachev) [inline include/communication/models/communication_description.h:19]: new line

File: include/communication/models/communication.h [hunk 29]
```diff
@@ -0,0 +1,48 @@
...
+                  const domain::ProtocolSpecification& protocolDescription,
+                  ICommunicationProtocol* protocol, const QString& name = QString(),
+                  QObject* parent = nullptr);
+    QString type() const override;
+    bool isConnected() const override;
+    int bytesReceived() const override;
+    int bytesSent() const override;
+
+public slots:
+    void setConnected(bool connected) override;
```
Comment (MishkaRogachev) [inline include/communication/models/communication.h:34]: What is the difference between setConnected and start/stop?

File: include/communication/models/link_specification.h [hunk 30]
```diff
@@ -0,0 +1,30 @@
...
+#define LINK_SPECIFICATION_H
+
+#include "communication_traits.h"
+
+#include <QObject>
+#include <QVariantMap>
+
+namespace md::domain
+{
+class LinkSpecification : public QObject
```
Comment (MishkaRogachev) [inline include/communication/models/link_specification.h]: No need in QObject if it just VO

File: include/communication/models/link_specification.h [hunk 31]
```diff
@@ -0,0 +1,30 @@
...
+    LinkSpecification(const QVariantMap& parameters, QObject* parent = nullptr);
+    LinkSpecification(const LinkSpecification& another, QObject* parent = nullptr);
+
+    QString type() const;
+    QVariant parameter(const QString& parameter) const;
+    QVariantMap parameters() const;
+
+private:
+    const QString m_type;
+    QVariantMap m_parameters;
```
Comment (MishkaRogachev) [inline include/communication/models/link_specification.h]: const

File: include/communication/models/protocol_specification.h [hunk 32]
```diff
@@ -0,0 +1,27 @@
+#ifndef PROTOCOL_SPECIFICATION_H
+#define PROTOCOL_SPECIFICATION_H
+
+#include <QObject>
+#include <QVariant>
+
+namespace md::domain
+{
+class ProtocolSpecification : public QObject
```
Comment (MishkaRogachev) [inline include/communication/models/protocol_specification.h]: QObject is overhead for VO

File: include/communication/models/protocol_specification.h [hunk 33]
```diff
@@ -0,0 +1,27 @@
...
+{
+    Q_OBJECT
+
+public:
+    ProtocolSpecification() = default;
+    ProtocolSpecification(const QString& name, QObject* parent = nullptr);
+    ProtocolSpecification(const ProtocolSpecification& another, QObject* parent = nullptr);
```
Comment (MishkaRogachev) [inline include/communication/models/protocol_specification.h]: copy-constructors generated by default for  VO

File: include/communication/service/communication_service.h [hunk 34]
```diff
@@ -0,0 +1,44 @@
+#ifndef COMMUNICATION_SERVICE_H
+#define COMMUNICATION_SERVICE_H
+
+#include "i_service.h"
```
Comment (MishkaRogachev) [inline include/communication/service/communication_service.h]: move to internal headers

File: include/communication/service/communication_service.h [hunk 35]
```diff
@@ -0,0 +1,44 @@
...
+#include <QObject>
+
+#include <QScopedPointer>
```
Comment (MishkaRogachev) [inline include/communication/service/communication_service.h]: unneeded

File: include/communication/service/communication_service.h [hunk 36]
```diff
@@ -0,0 +1,44 @@
...
+class CommunicationService
+    : public QObject
+    , public md::domain::IService
+{
+    Q_OBJECT
+
+public:
+    CommunicationService(const QString& fileName);
+
+    void registerProtocol(const QString& name, data_source::ICommunicationProtocol* protocol);
```
Comment (MishkaRogachev) [inline include/communication/service/communication_service.h:26]: const QString& id

**Review by MishkaRogachev — CHANGES_REQUESTED** [review]

File: include/communication/models/communication.h (hunk 29 above)
Comment (nnhw) [reply include/communication/models/communication.h:34]: I don't know how exactly I should implement them, but for now start/stop is for creation/destruction and threading; 
connect/disconnect is for opening/closing sockets in the future

File: include/communication/service/communication_service.h (hunk 36 above)
Comment (nnhw) [reply include/communication/service/communication_service.h:26]: At this time, protocol specification holds name, not ID, as Jerom. We should not change it in one place only

File: include/communication/service/communication_service.h (hunk 34 above)
Comment (nnhw) [reply include/communication/service/communication_service.h]: Do u mean to cpp? But we inherit from IService and we need this in the header

Comment (MishkaRogachev) [reply include/communication/service/communication_service.h]: No, just place include with other internal includes

File: include/communication/service/communication_service.h (hunk 36 above)
Comment (MishkaRogachev) [reply include/communication/service/communication_service.h:26]: In futer will be need both id and name

File: include/communication/service/communication_service.h (hunk 34 above)
Comment (nnhw) [reply include/communication/service/communication_service.h]: I thought that we place parent class headlers first, as in here - https://github.com/Midgrad/Dreka/wiki/Coding-Styles 

But ok

**Pull request merged**